README.md
Cabot Slack Plugin
=====

Based on: https://github.com/bonniejools/cabot-alert-hipchat

This is an alert plugin for the cabot service monitoring tool. It allows you to alert users by their user handle in a slack channel.

## Installation

Enter the cabot virtual environment.
```
    $ pip install cabot-alert-slack
    $ foreman stop
```

or

```
    $ pip install git+git://github.com/lblasc/cabot-alert-slack.git
    $ foreman stop
```

Edit `conf/*.env`.

```
CABOT_PLUGINS_ENABLED=cabot_alert_slack==0.6
...
SLACK_ALERT_CHANNEL=channel_name_without_hash
SLACK_WEBHOOK_URL=url_of_your_webhook_integration_from_slack
SLACK_ICON_URL=http://lorempixel.com/40/40/
```

Add cabot_alert_slack to the installed apps in settings.py
```
    $ foreman run python manage.py syncdb
    $ foreman start
```


CHANGES
Version 0.8.3
-------------

* Add serialize method

Version 0.8.2
-------------

* Fix `short` attachment property by using boolean instead of string
   - This allows to use the plugin with Mattermost
* Fix link to Jenkins jobs with multiple Jenkins backends
   - This was added in cabot 0.11.0

Version 0.8.1
-------------

* Fix slack @mentions not linking correctly
  - Note: It still doesn't work with app webhooks, potentially due to a slack bug
* Only show slack buttons if SLACK_INTERACTIVE_MESSAGES is set to true
  - Some custom config is required on slacks end, so doesn't make sense to show it by default


Version 0.8.0
-------------

* Add slack acknowledge button and callback

Version 0.7.0
-------------

* Add migrations to work with Cabot >=0.7


cabot_alert_slack/tests/test_slack.py
from cabot.cabotapp.tests.tests_basic import LocalTestCase
from mock import Mock, patch

from cabot.cabotapp.models import UserProfile, Service
from cabot_alert_slack import models
from cabot.cabotapp.alert import update_alert_plugins

class TestSlackAlerts(LocalTestCase):
    def setUp(self):
        super(TestSlackAlerts, self).setUp()

        self.user_profile = UserProfile(user=self.user)
        self.user_profile.save()
        self.slack_user_data = models.SlackAlertUserData.objects.create(
            slack_alias = "test_user_slack_alias",
            user = self.user_profile,
            title=models.SlackAlertUserData.name,
            )
        self.slack_user_data.save()

        self.service.users_to_notify.add(self.user)

        update_alert_plugins()
        self.slack_plugin = models.SlackAlert.objects.get(title=models.SlackAlert.name)
        self.service.alerts.add(self.slack_plugin)
        self.service.save()
        self.service.update_status()

    def test_users_to_notify(self):
        self.assertEqual(self.service.users_to_notify.all().count(), 1)
        self.assertEqual(self.service.users_to_notify.get(pk=1).username, self.user.username)

    @patch('cabot_alert_slack.models.SlackAlert._send_slack_alert')
    def test_normal_alert(self, fake_slack_alert):
        self.service.overall_status = Service.PASSING_STATUS
        self.service.old_overall_status = Service.ERROR_STATUS
        self.service.save()
        self.service.alert()
        fake_slack_alert.assert_called_with(u'Service Service is back to normal: http://localhost/service/1/. @test_user_slack_alias', color='green', sender='Cabot/Service')

    @patch('cabot_alert_slack.models.SlackAlert._send_slack_alert')
    def test_failure_alert(self, fake_slack_alert):
        # Most recent failed
        self.service.overall_status = Service.CALCULATED_FAILING_STATUS
        self.service.old_overall_status = Service.PASSING_STATUS
        self.service.save()
        self.service.alert()
        fake_slack_alert.assert_called_with(u'Service Service reporting failing status: http://localhost/service/1/. Checks failing: @test_user_slack_alias', color='red', sender='Cabot/Service')


cabot_alert_slack/views.py
from django.views.decorators.csrf import csrf_exempt
from django.http import JsonResponse
from .models import SlackAlertUserData

from cabot.cabotapp.models import Service
import json
import re


@csrf_exempt
def slack_message_callback(request):
    payload = json.loads(request.POST['payload'])
    service_id = re.match('acknowledge_(\d+)', payload['callback_id']).groups()[0]

    slack_alias = payload['user']['name']
    # .user.user to go through user_profile -> user...
    user = SlackAlertUserData.objects.get(slack_alias=slack_alias).user.user

    service = Service.objects.get(pk=service_id)
    service.acknowledge_alert(user)

    message = payload['original_message']
    # Strip out button
    for attach in message['attachments']:
        if int(attach['id']) != int(payload['attachment_id']):
            continue
        attach['actions'] = []

    message['attachments'].append({
        'text': 'Acknowledged by @{}'.format(slack_alias),
        'color': 'warning',
    })

    return JsonResponse(message)


cabot_alert_slack/migrations/0001_initial.py
# -*- coding: utf-8 -*-
# Generated by Django 1.9.12 on 2017-03-21 15:14
from __future__ import unicode_literals

from django.db import migrations, models
import django.db.models.deletion


class Migration(migrations.Migration):

    initial = True

    dependencies = [
        ('cabotapp', '0003_auto_20170201_1045'),
    ]

    operations = [
        migrations.CreateModel(
            name='SlackAlert',
            fields=[
                ('alertplugin_ptr', models.OneToOneField(auto_created=True, on_delete=django.db.models.deletion.CASCADE, parent_link=True, primary_key=True, serialize=False, to='cabotapp.AlertPlugin')),
            ],
            options={
                'abstract': False,
            },
            bases=('cabotapp.alertplugin',),
        ),
        migrations.CreateModel(
            name='SlackAlertUserData',
            fields=[
                ('alertpluginuserdata_ptr', models.OneToOneField(auto_created=True, on_delete=django.db.models.deletion.CASCADE, parent_link=True, primary_key=True, serialize=False, to='cabotapp.AlertPluginUserData')),
                ('slack_alias', models.CharField(blank=True, max_length=50)),
            ],
            options={
                'abstract': False,
            },
            bases=('cabotapp.alertpluginuserdata',),
        ),
    ]


cabot_alert_slack/models.py
from django.db import models
from cabot.cabotapp.alert import AlertPlugin, AlertPluginUserData

from os import environ as env

from django.conf import settings
from django.core.urlresolvers import reverse
from django.template import Context, Template

import requests
import json

slack_template = """
Service {{ service.name }} {% if service.overall_status == service.PASSING_STATUS %}*is back to normal*{% else %}reporting *{{ service.overall_status }}* status{% endif %}: {{ scheme }}://{{ host }}{% url 'service' pk=service.id %} \
{% if alert %}{% for alias in users %} @{{ alias }}{% endfor %}{% endif %}\

{% if service.overall_status != service.PASSING_STATUS %}Checks failing:\
{% for check in service.all_failing_checks %}\
    {% if check.check_category == 'Jenkins check' %}\
        {% if check.last_result.error %}\
            - {{ check.name }} ({{ check.last_result.error|safe }}) {{check.jenkins_config.jenkins_api}}job/{{ check.name }}/{{ check.last_result.job_number }}/console
        {% else %}\
            - {{ check.name }} {{check.jenkins_config.jenkins_api}}/job/{{ check.name }}/{{check.last_result.job_number}}/console
        {% endif %}\
    {% else %}
        - {{ check.name }} {% if check.last_result.error %} ({{ check.last_result.error|safe }}){% endif %}
    {% endif %}\
{% endfor %}\
{% endif %}\
"""

# This provides the slack alias for each user. Each object corresponds to a User
class SlackAlert(AlertPlugin):
    name = "Slack"
    author = "Luka Blaskovic"

    def send_alert(self, service, users, duty_officers):
        alert = True
        slack_aliases = []
        users = list(users) + list(duty_officers)

        slack_aliases = [u.slack_alias for u in SlackAlertUserData.objects.filter(user__user__in=users)]

        if service.overall_status == service.WARNING_STATUS:
            alert = False  # Don't alert at all for WARNING
        if service.overall_status == service.ERROR_STATUS:
            if service.old_overall_status in (service.ERROR_STATUS, service.ERROR_STATUS):
                alert = False  # Don't alert repeatedly for ERROR
        if service.overall_status == service.PASSING_STATUS:
            color = 'good'
            if service.old_overall_status == service.WARNING_STATUS:
                alert = False  # Don't alert for recovery from WARNING status
        else:
            color = 'danger'

        c = Context({
            'service': service,
            'users': slack_aliases,
            'host': settings.WWW_HTTP_HOST,
            'scheme': settings.WWW_SCHEME,
            'alert': alert,
        })
        message = Template(slack_template).render(c)
        self._send_slack_alert(message, service, color=color, sender='Cabot')

    def _send_slack_alert(self, message, service, color='good', sender='Cabot'):

        channel = '#' + env.get('SLACK_ALERT_CHANNEL')
        url = env.get('SLACK_WEBHOOK_URL')
        icon_url = env.get('SLACK_ICON_URL')

        actions = []
        if env.get('SLACK_INTERACTIVE_MESSAGES', False) and service.overall_status != service.PASSING_STATUS:
            actions.append({
                "name": "acknowledge",
                "text": "Acknowledge",
                "type": "button",
                "value": "acknowledge",
            })

        # TODO: handle color
        resp = requests.post(url, data=json.dumps({
            'channel': channel,
            'username': sender[:15],
            'icon_url': icon_url,
            'link_names': 1,
            'attachments': [{
                'title': service.name,
                'text': message,
                'color': color,
                'mrkdwn_in': ["text"],
                'fields': [{
                    'title': 'status',
                    'value': service.overall_status,
                    'short': True
                    }, {
                    'title': 'old status',
                    'value': service.old_overall_status,
                    'short': True
                    }
                ],
                'callback_id': "acknowledge_{}".format(service.id),
                'actions': actions
            }]
        }))

class SlackAlertUserData(AlertPluginUserData):
    name = "Slack Plugin"
    slack_alias = models.CharField(max_length=50, blank=True)

    def serialize(self):
        return {
            "slack_alias": self.slack_alias
        }